--- Entity.pm.orig	Tue Feb 24 03:04:02 1998
+++ Entity.pm	Thu Feb 26 03:05:07 1998
@@ -525,7 +525,7 @@
 sub build {
     my ($self, @paramlist) = @_;
     my %params = @paramlist;
-    my ($field, $filename, $boundary);
+    my ($charset, $field, $filename, $boundary);
 
     # Create a new entity, if needed:
     ref($self) or $self = $self->new;
@@ -543,9 +543,12 @@
 
     # Get recommended filename, allowing explicit no-value value:
     $filename = (exists($params{Filename}) 
-		 ? $params{Filename} : ($params{Path} || ''));
+		 ? defined($params{Filename}) ? $params{Filename} : ''
+                 : ($params{Path} || ''));
     $filename =~ s{^.*/}{}g;        # nuke path info    
 
+    $charset = $params{Charset};
+ 
     # Type-check sanity:
     if ($type =~ m{^(multipart|message)/}) {
 	($encoding =~ /^(|7bit|8bit|binary|-suggest)$/i) 
@@ -597,6 +600,7 @@
     # Add content-type field:
     $field = new Mail::Field 'Content_type';         # not a typo :-(
     $field->type($type);
+    $field->charset($charset)    if ($charset);
     $field->name($filename)      if ($filename ne '');
     $field->boundary($boundary)  if (defined($boundary));
     $head->replace('Content-type', $field->stringify);
